moodle/podcontroller.go

```diff
@@ -0,0 +1,242 @@
+func (c *PodController) enqueuePod(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	fmt.Printf("PodController.go     : enqueuePod, adding a pod key %s\n", key)
+	c.podqueue.AddRateLimited(key)
+}
+func (c *PodController) handlePod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key))
+		return fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key)
+	}
+	pod, err := c.kubeclientset.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return fmt.Errorf("Could not find the pod for pod name : %s, namespace %s", name, namespace)
+	}
+	ownerReferences := pod.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : Pod does not have any owner references.")
+		return nil
+	}
+	ownerReferenceName := ownerReferences[0].Name
+
+	replicaSet, err := c.kubeclientset.AppsV1().ReplicaSets(namespace).Get(ownerReferenceName, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return nil
+	}
+	ownerReferences = replicaSet.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : replicaset does not have a ny owner references.")
+		return nil
+	}
+	ownerReferenceName = ownerReferences[0].Name
+
+	deployment, err := c.kubeclientset.AppsV1().Deployments(namespace).Get(ownerReferenceName, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the deployment, skipping")
+		return nil
+	}
+	ownerReferences = deployment.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : deployment does not have any owner references. Cannot be a moodle.")
+		return nil
+	}
+
+	if ownerReferences[0].Kind != "Moodle" {
+		// fmt.Println("PodController.go     : The pod event is not a Moodle instance. Skipping!")
+		return nil
+	}
+
+	moodle, err := c.moodleLister.Moodles(namespace).Get(ownerReferences[0].Name)
+	if err != nil {
+		return fmt.Errorf("PodController.go: Could not find the Moodle instance : %s", ownerReferences[0].Name)
+	}
+	if !c.waitForPod(constants.TIMEOUT, name, namespace) {
+		return fmt.Errorf("PodController.go: The Moodle Pod failed to start running")
+	}
+	erredPlugins := c.util.EnsurePluginsInstalled(moodle, name, namespace, constants.PLUGIN_MAP)
+	if len(erredPlugins) > 0 {
+		return fmt.Errorf("PodController.go: Some plugins failed to install. Cannot ensure state. %v", erredPlugins)
+	}
+	return nil
+}
+
+func (c *PodController) waitForPod(timeout int, podname, namespace string) bool {
```

Comment:
Is this copy of the method from moodle.go? Is so, we should refactor this method into utils and make calls to it from moodle.go and from here.

Reply:
Ok I am working on this